lib/Dialect/Datapath/DatapathFolds.cpp

```diff
@@ -0,0 +1,260 @@
+// Compress Operation
+//===----------------------------------------------------------------------===//
+// Check that all compressor results are included in this list of operands
+// If not we must take care as manipulating compressor results independently
+// could easily introduce a non-equivalent representation.
+static bool areAllCompressorResultsSummed(ValueRange compressResults,
+                                          ValueRange operands) {
+  for (auto result : compressResults) {
+    if (!llvm::is_contained(operands, result))
+      return false;
+  }
+  return true;
+}
+
+struct FoldCompressIntoCompress
+    : public OpRewritePattern<datapath::CompressOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  // compress(compress(a,b,c), add(e,f)) -> compress(a,b,c,e,f)
+  LogicalResult matchAndRewrite(datapath::CompressOp compOp,
+                                PatternRewriter &rewriter) const override {
+    auto operands = compOp.getOperands();
+    llvm::SmallSetVector<Value, 8> processedCompressorResults;
+    SmallVector<Value, 8> newCompressOperands;
+
+    for (Value operand : operands) {
+
+      // Skip if already processed this compressor
+      if (processedCompressorResults.contains(operand))
+        continue;
+
+      // If the operand has multiple uses, we do not fold it into a compress
+      // operation, so we treat it as a regular operand to maintain sharing.
+      if (!operand.hasOneUse()) {
+        newCompressOperands.push_back(operand);
+        continue;
+      }
+
+      // Found a compress op - add its operands to our new list
+      if (auto compressOp = operand.getDefiningOp<datapath::CompressOp>()) {
+
+        // Check that all results of the compressor are summed in this add
+        if (!areAllCompressorResultsSummed(compressOp.getResults(), operands))
+          return failure();
+
+        llvm::append_range(newCompressOperands, compressOp.getOperands());
+        // Only process each compressor once as multiple operands will point
+        // to the same defining operation
+        processedCompressorResults.insert(compressOp.getResults().begin(),
+                                          compressOp.getResults().end());
+        continue;
+      }
+
+      if (auto addOp = operand.getDefiningOp<comb::AddOp>()) {
+        llvm::append_range(newCompressOperands, addOp.getOperands());
+        continue;
+      }
+
+      // Regular operand - just add it to our list
+      newCompressOperands.push_back(operand);
+    }
+
+    // If unable to collect more operands then this pattern doesn't apply
+    if (newCompressOperands.size() <= compOp.getNumOperands())
+      return failure();
+
+    // Create a new CompressOp with all collected operands
+    rewriter.replaceOpWithNewOp<datapath::CompressOp>(
+        compOp, newCompressOperands, compOp.getNumResults());
+    return success();
+  }
+};
+
+struct FoldAddIntoCompress : public OpRewritePattern<comb::AddOp> {
```

Comment:
Nit: it often helps readability a lot to add a typical example of an application of a canonicalization to the docsting of the pattern struct.